distributed/deploy/spec.py

```diff
@@ -306,7 +306,10 @@ def _update_worker_status(self, op, msg):
             name = self.scheduler_info["workers"][msg]["name"]
 
             def f():
-                if name in self.workers and msg not in self.scheduler_info:
+                if (
+                    msg in [w.worker_address for w in self.workers.values()]
```

Comment:
The `Worker` and `Nanny` classes have the `worker_address` attribute, but it's not guaranteed to be present.  dask-jobqueue `Job` objects don't know the address of the worker at all for example.

Reply:
I wonder if it makes sense to close the Worker object here if no other worker has shown up with the same name?  

This seems error prone though, because we're likely to get the "removed" message before the new worker has a chance to spin up.

Reply:
Thoughts on this @jrbourbeau ?

Reply:
Maybe something like the following?

```python
if (name in self.workers and 
    msg not in self.scheduler_info["workers"] and
    not any(d["name"] == name for d in self.scheduler_info["workers"].values()
):
```

Reply:
This works if the new worker has already spun up. If the new worker has not spun up yet then this will delete the old worker (in line with https://github.com/dask/distributed/pull/2990#issuecomment-527218288), but the new workers then don't show up in `self.workers` (still not sure why yet). The new workers exist and are functional (e.g. you can compute things on them), but aren't in `self.workers`. 

```python
from time import sleep
from distributed import Client, wait
import dask


def inc(x):
    return x + 1


if __name__ == "__main__":

    with dask.config.set({"distributed.deploy.lost-worker-timeout": "1ms"}):
        with Client(n_workers=1) as client:
            print(f"client.cluster.workers = {client.cluster.workers}")
            client.restart()
            for _ in range(5):
                print(f"client.cluster.workers = {client.cluster.workers}")
                sleep(0.5)

            future = client.submit(inc, 1)
            wait(future)
            print(f"future.result() = {future.result()}")
            print(f"client.who_has() = {client.who_has()}")
            print(f"client.cluster.workers = {client.cluster.workers}")

```

Reply:
> If the new worker has not spun up yet then this will delete the old worker 

I think that this is what the current delay is for?  We wait a modest time to allow the worker to come back.  If it does, then we don't delete it.

> but the new workers then don't show up in self.workers

From the cluster manager's perspective, there is no new worker.  In the `Client()` case the objects in `self.worker` are `Nanny` objects, which aren't dying or coming back.  They've been the same the entire time.

Reply:
I guess I'm talking about when we wait a very small amount of time for the worker to come back (e.g. 1ms). So we're checking if the new worker has appeared before the `"add"` message (from `restart`-ing the workers) has been received by the cluster manager. 

In this case, the new worker hasn't appeared yet, so we remove the `Nanny` object from `self.workers`. But later the `"add"` message is received and the new worker appears. However the corresponding `Nanny` is not added back to `self.workers`.

Does it make sense to set a minimum allowed value for `lost-worker-timeout`?

Reply:
> Does it make sense to set a minimum allowed value for lost-worker-timeout?

Sorry for going dark here.

I don't think so.  I think that our default should definitely be large enough so that a worker is likely to restart in this time (I think that it is).  But that if a user wants to claim that their workers restart faster, then we should let them, even if they're wrong.  In practice I doubt that many will modify this configuration value.

Reply:
So without thinking too hard about this I think that we should do this

```python
if (name in self.workers and 
    msg not in self.scheduler_info["workers"] and
    not any(d["name"] == name for d in self.scheduler_info["workers"].values()
):
```

I think that your concern here 

> This works if the new worker has already spun up. If the new worker has not spun up yet then this will delete the old worker

Is accurate, but that we need to assume that the delay will handle this.

There are probably also other solutions here, but I think that this solution is decent, and improves things.

Reply:
I agree with your points about letting users claim their workers start quickly and also that the fix here is net progress. I've updated the logic in `_update_worker_status` accordingly